Model/Datasource/Database/MysqlExtended.php

```diff
@@ -99,4 +116,169 @@ public function column($real) {
+		return parent::execute($sql, $options, $params);
+	}
+
+	/**
+	 * try to re-execute again, maybe
+	 * based on the query, exception, etc
+	 *
+	 * @param PDOException $e
+	 * @param string $sql SQL statement
+	 * @param array $options The options for executing the query.
+	 * @param array $params values to be bound to the query.
+	 * @return mixed Resource or object representing the result set, or false on failure
+	 */
+	public function tryExecuteAgain($e, $sql, $options, $params) {
+		if (!$this->shouldWeExecuteAgain($e)) {
+			$this->log('MysqlExtended: Did not retry query THROWING: ' . $sql);
+			throw $e;
+		}
+		$this->tryExecuteAgainSleep($e);
+		$this->tryExecuteAgainAddRepeat($e);
+
+		$this->log('MysqlExtended: Attempting to retry query: ' . $sql);
+		$return = $this->execute($sql, $options, $params);
+		$this->log('MysqlExtended: Successful retry of query: ' . $sql);
+
+		return $return;
+	}
+
+	/**
+	 * should we try to re-execute again?
+	 * based on the query, exception, etc
+	 *
+	 * @param PDOException $e
+	 * @return boolean
+	 */
+	public function shouldWeExecuteAgain($e) {
+		if (!($e instanceof PDOException)) {
+			return false;
+		}
+		if (!isset($e->queryString)) {
+			return false;
+		}
+		if (!$this->shouldWeExecuteAgainAllowedMessage($e)) {
+			return false;
+		}
+		if (!$this->shouldWeExecuteAgainAllowedRepeat($e)) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * should we try to re-execute again?
+	 * based on the query message
+	 *
+	 * @param PDOException $e
+	 * @return boolean
+	 */
+	public function shouldWeExecuteAgainAllowedMessage($e) {
+		// TODO consider limiting to specific error codes: 1213, 1205
+		return (strpos($e->getMessage(), 'try restarting transaction') !== false);
+	}
+
+	/**
+	 * should we try to re-execute again?
+	 * based on the already repeated X times
+	 *
+	 * @param PDOException $e
+	 * @return boolean
+	 */
+	public function shouldWeExecuteAgainAllowedRepeat($e) {
+		if (!isset($this->_tryExecuteAgain[$e->queryString])) {
+			$this->_tryExecuteAgain[$e->queryString] = 0;
+		}
+		return ($this->_tryExecuteAgain[$e->queryString] < $this->_tryExecuteAgainConfig['limit']);
+	}
+
+	/**
+	 * add 1 to the count of retries for this query
+	 *
+	 * @param PDOException $e
+	 * @return boolean
+	 */
+	public function tryExecuteAgainAddRepeat($e) {
+		if (!isset($this->_tryExecuteAgain[$e->queryString])) {
+			$this->_tryExecuteAgain[$e->queryString] = 1;
+			return;
+		}
+		$this->_tryExecuteAgain[$e->queryString]++;
+	}
+
+	/**
+	 * add 1 to the count of retries for this query
+	 *
+	 * @return boolean
+	 */
+	public function tryExecuteAgainUnsetRepeat() {
+		unset($this->_tryExecuteAgain);
+	}
+
+	/**
+	 * add 1 to the count of retries for this query
+	 *
+	 * @param PDOException $e
+	 * @return boolean
+	 */
+	public function tryExecuteAgainSleep($e) {
+		if (strpos($e->getMessage(), '1205') !== false) {
+			// error 1205 is a long timeout, so we sleep longer: 100 ms
+			usleep(10000);
```

Comment:
there isn't - i can safely be removed... but I would like to keep the sleep timeout a split out function in case we need to tweak it or make it more dynamic... (I think I was future-thinking too much there)